src/mcpwm.rs

```diff
@@ -0,0 +1,527 @@
+                    cmpr_a: config.duty_a,
+                    cmpr_b: config.duty_b,
+                    duty_mode: config.duty_mode.into(),
+                    counter_mode: config.counter_mode.into(),
+                },
+            )
+        })?;
+
+        let pin_a: Option<PA> = pin_a.into();
+        let pin_b: Option<PB> = pin_b.into();
+
+        if let Some(pin_a) = &pin_a {
+            let io_signal = O::signal_a();
+            esp!(unsafe { esp_idf_sys::mcpwm_gpio_init(U::unit(), io_signal, pin_a.pin()) })?;
+        }
+
+        if let Some(pin_b) = &pin_b {
+            let io_signal = O::signal_b();
+            esp!(unsafe { esp_idf_sys::mcpwm_gpio_init(U::unit(), io_signal, pin_b.pin()) })?;
+        }
+
+        Ok(Self {
+            _instance: operator,
+            _unit: U::default(),
+            _mcpwm_module: mcpwm_module,
+            _pin_a: pin_a,
+            _pin_b: pin_b,
+        })
+    }
+
+    pub fn get_duty_a(&self) -> Duty {
+        unsafe {
+            esp_idf_sys::mcpwm_get_duty(
+                U::unit(),
+                O::timer(),
+                esp_idf_sys::mcpwm_generator_t_MCPWM_GEN_A,
+            )
+        }
+    }
+
+    pub fn get_duty_b(&self) -> Duty {
+        unsafe {
+            esp_idf_sys::mcpwm_get_duty(
+                U::unit(),
+                O::timer(),
+                esp_idf_sys::mcpwm_generator_t_MCPWM_GEN_B,
+            )
+        }
+    }
+
+    /// Set duty as percentage between 0.0 and 100.0
+    pub fn set_duty_a(&mut self, duty: Duty) -> Result<(), EspError> {
+        unsafe {
+            esp!(esp_idf_sys::mcpwm_set_duty(
+                U::unit(),
+                O::timer(),
+                esp_idf_sys::mcpwm_generator_t_MCPWM_GEN_A,
+                duty
+            ))
+        }
+    }
+
+    pub fn set_duty_b(&mut self, duty: Duty) -> Result<(), EspError> {
+        unsafe {
+            esp!(esp_idf_sys::mcpwm_set_duty(
+                U::unit(),
+                O::timer(),
+                esp_idf_sys::mcpwm_generator_t_MCPWM_GEN_B,
+                duty
+            ))
+        }
+    }
+
+    pub fn set_frequency(frequency: u32) -> Result<(), EspError> {
+        unsafe {
+            esp!(esp_idf_sys::mcpwm_set_frequency(
+                U::unit(),
+                O::timer(),
+                frequency
+            ))
+        }
+    }
+    pub fn get_frequency() -> u32 {
+        unsafe { esp_idf_sys::mcpwm_get_frequency(U::unit(), O::timer()) }
+    }
+}
+
+pub enum Generator {
+    A,
+    B,
+}
+
+// TODO: would we like to also implement PwmPin for something like Foo<&Operator, G: Generator>
+// this would allow things like:
+// let operator = Operator::new(...);
+// let gen_a = operator.generator_a_as_pwm_pin();
+// let gen_b = operator.generator_b_as_pwm_pin();
+// function_expecting_pwm_pin(&gen_a);
+// gen_a.set_duty()
+// gen_a.set_duty()
+impl<U: Unit, O: HwOperator<U>, M: Borrow<Mcpwm<U>>, PA: OutputPin, PB: OutputPin>
```

Comment:
This would be nice! But you can leave it as TODO for now, if you don't want to add this immediatly.